lucille-plugins/lucille-pinecone/src/main/java/com/kmwllc/lucille/pinecone/indexer/PineconeIndexer.java

```diff
@@ -95,52 +97,60 @@ protected void sendToIndex(List<Document> documents) throws IndexerException {
-              Collectors.toList());
-        } else {
-          uploadList = uploadList.stream().filter(other -> !other.getId().equals(doc.getId())).collect(Collectors.toList());
-        }
+        // always add to deleteList;
+        // in the case where doc is in uploadList and is now marked for deletion, then remove that document in uploadList
+        uploadList.removeIf(x -> x.getId().equals(doc.getId()));
         deleteList.add(doc);
       } else {
-        // if doc is already in deleteList, then continue adding to uploadList as uploading is performed after deletion
-        // note: updating/deleting a non-existent ID will return an OK response
+        // always add non deletion documents to uploadList;
+        // in the case that doc is also in deleteList, we remove it from the deleteList to avoid sending unnecessary deletion request
+        deleteList.removeIf(x -> x.getId().equals(doc.getId()));
         uploadList.add(doc);
       }
     }
 
     // if there exists documents to delete
     if (!deleteList.isEmpty()) {
+      validateDeleteRequirements(deleteList.size());
       if (namespaces != null) {
         for (Map.Entry<String, Object> entry : namespaces.entrySet()) {
-          deleteDocuments(deleteList, addDeletionPrefix, entry.getKey());
+          deleteDocuments(deleteList, entry.getKey());
         }
       } else {
-        deleteDocuments(deleteList, addDeletionPrefix,"default");
+        deleteDocuments(deleteList, PineconeManager.getDefaultNamespace());
       }
     }
 
     // if there exists documents to upload
     if (!uploadList.isEmpty()) {
       // check that both namespaces and defaultEmbeddingField is not null only when uploading
-      validateUploadRequirements();
+      validateUploadRequirements(uploadList.size());
       if (namespaces != null) {
         for (Map.Entry<String, Object> entry : namespaces.entrySet()) {
           uploadDocuments(uploadList, (String) entry.getValue(), entry.getKey());
         }
       } else {
-        uploadDocuments(uploadList, defaultEmbeddingField, "default");
+        uploadDocuments(uploadList, defaultEmbeddingField, PineconeManager.getDefaultNamespace());
       }
     }
   }
 
-  private void validateUploadRequirements() {
+  private void validateUploadRequirements(int uploadListSize) throws IndexerException {
     if (namespaces == null && defaultEmbeddingField == null) {
       throw new IllegalArgumentException(
           "At least one of a defaultEmbeddingField or a non-empty namespaces mapping is required when uploading documents.");
     }
+    if (uploadListSize > MAX_PINECONE_BATCH_SIZE) {
+      throw new IndexerException(
+          "Pinecone maximum batch size is " + MAX_PINECONE_BATCH_SIZE + ", lower batchSize in indexer configs.");
+    }
+  }
+
+  private void validateDeleteRequirements(int deleteListSize) throws IndexerException {
+    if (deleteListSize > MAX_PINECONE_BATCH_SIZE) {
```

Comment:
This code can be abstracted to a new method validateBatchSizeRequirements() or similar. This could then be called in validateUploadRequirements and be called directly on deleteList.size() to avoid copied code.